### PR TITLE
[SEARCH-174] wand go brrr

### PR DESCRIPTION
**Pull request by MBkkt**
### Scope & Purpose

Enable speedup term query with optimizeTopK in 10-1000 times

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [x] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/1249
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/Aql/IResearchViewExecutor.h [hunk 1]
```diff
@@ -112,6 +114,10 @@ class IResearchViewExecutorInfos {
     return _filterOptimization;
   }
 
+  iresearch::IResearchOptimizeTopK const& optimizeTopK() const noexcept {
```
Comment (Dronplane) [inline arangod/Aql/IResearchViewExecutor.h:117]: #ifdef USE_ENTERPRISE

**Review by gnusi — APPROVED** [review]

File: arangod/Aql/IResearchViewExecutor.tpp [hunk 2]
```diff
@@ -1147,8 +1163,16 @@ bool IResearchViewHeapSortExecutor<ExecutionTraits>::fillBufferInternal(
...
           scr = &irs::score::kNoScore;
           numScores = 0;
+          threshold = nullptr;
         } else {
           numScores = scores.size();
+          auto* new_threshold =
+              irs::get_mutable<irs::score_threshold>(itr.get());
+          if (threshold != nullptr) {
+            TRI_ASSERT(new_threshold->value == 0.f);
+            new_threshold->value = threshold->value;
```
Comment (Dronplane) [inline arangod/Aql/IResearchViewExecutor.tpp:1173]: threshold is invalid pointer here

Comment (MBkkt) [reply arangod/Aql/IResearchViewExecutor.tpp]: Fixed

File: arangod/Aql/IResearchViewNode.cpp [hunk 3]
```diff
@@ -720,6 +720,28 @@ std::shared_ptr<SearchMeta const> getMeta(
   return basics::downCast<Search>(*view).meta();
 }
 
+#ifdef USE_ENTERPRISE
+
+IResearchOptimizeTopK const& optimizeTopK(
+    std::shared_ptr<SearchMeta const> const& meta,
+    std::shared_ptr<LogicalView const> const& view) {
+  TRI_ASSERT(!ServerState::instance()->isCoordinator());
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply arangod/Aql/IResearchViewNode.cpp]: Outdated

File: arangod/IResearch/Search.cpp [hunk 4]
```diff
@@ -264,6 +264,11 @@ std::string checkFieldsDifferentCollections(
 
 std::string check(SearchMeta const& search,
                   IResearchInvertedIndexMeta const& index) {
+#ifdef USE_ENTERPRISE
+  if (search.optimizeTopK != index._optimizeTopK) {
+    return "index optimize topK mismatches view optimize topK";
```
Comment (Dronplane) [inline arangod/IResearch/Search.cpp:269]: NIT  to match actual parameter name.
```suggestion
    return "index optimizeTopK mismatches view optimizeTopK";
```

Comment (MBkkt) [reply arangod/IResearch/Search.cpp:269]: I made it consistent with other messages

**Review by Dronplane — APPROVED** [review]
LGTM

**Comment by Dronplane** [comment]
Tests blue except known failure

**Pull request merged**